portfolio/src/main/java/com/google/sps/servlets/HomeServlet.java

```diff
@@ -0,0 +1,72 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.*;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/login")
+public class HomeServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html");
+    PrintWriter out = response.getWriter();
+    UserService userService = UserServiceFactory.getUserService();
+
+    // If user is not logged in, show a login form (could also redirect to a login page)
+    if (!userService.isUserLoggedIn()) {
+      String loginUrl = userService.createLoginURL("/loginn");
+      out.println("<p>Login <a href=\"" + loginUrl + "\">here</a>.</p>");
+      return;
+    }
+
+    // If user has not set a nickname, redirect to nickname page
+    String nickname = getUserNickname(userService.getCurrentUser().getUserId());
+    if (nickname == null) {
+      response.sendRedirect("/nickname");
+      return;
+    }
+
+    // User is logged in and has a nickname, so the request can proceed
+    String logoutUrl = userService.createLogoutURL("/login");
+    out.println("<h1>Login</h1>");
+    out.println("<p>Hello " + nickname + "!</p>");
+    out.println("<p>Logout <a href=\"" + logoutUrl + "\">here</a>.</p>");
+    out.println("<p>Change your nickname <a href=\"/nickname\">here</a>.</p>");
```

Comment:
I think it is generally not a good idea to create HTML on the server if it can be helped. IMO all the HTML should be as close to each other as possible. Ideally in the .html file. Is it possible to return the logoutUrl or loginUrl to the server and have the html/js do this?

What do you think Matt?

Reply:
I will do that

Reply:
Agreed!  It looks like you have a JSON API between the frontend and the servlets.  I would *always* pass structured data (JSON) back from the servlets, and then choose how to render it to HTML in your JavaScript.  In a "real" application, I would extend that to include the login link you provide above – provide a servlet to let the frontend figure out whether someone is logged in, and let the frontend create the login link – but it's not a big deal for something so small and probably not worth the complication.